Sources/Networking/Modifiers/Interceptors/Authorization/AuthorizationTokenInterceptor.swift

```diff
@@ -0,0 +1,110 @@
+    
+    public init(authorizationManager: AuthorizationManaging) {
+        self.authorizationManager = authorizationManager
+    }
+}
+
+// MARK: - RequestInterceptor conformation
+public extension AuthorizationTokenInterceptor {
+    func adapt(_ request: URLRequest, for endpointRequest: EndpointRequest) async throws -> URLRequest {
+        guard endpointRequest.endpoint.isAuthenticationRequired else {
+            return request
+        }
+        
+        do {
+            return try await authorizationManager.authorizeRequest(request)
+        } catch {
+            /// If authorization fails due to expiredAccessToken we should perform refresh and then retry the request authorization again.
+            guard case AuthorizationError.expiredAccessToken = error else {
+                throw error
+            }
+            
+            try await performRefresh()
+            
+            return try await authorizationManager.authorizeRequest(request)
+        }
+    }
+    
+    func process(_ response: Response, with urlRequest: URLRequest, for endpointRequest: EndpointRequest) async throws -> Response {
+        guard let httpResponse = response.response as? HTTPURLResponse else {
+            throw NetworkError.noStatusCode(response: response)
+        }
+        
+        /// If request fails due to unauthorized but required valid authorization perform refresh before returning the response.
+        guard
+            httpResponse.statusCode == 401,
+            endpointRequest.endpoint.isAuthenticationRequired
+        else {
+            return response
+        }
+        
+        try await performRefresh()
+        
+        return response
+    }
+    
+    func process(_ error: Error, for endpointRequest: EndpointRequest) async -> Error {
+        error
+    }
+}
+
+// MARK: Private methods
+private extension AuthorizationTokenInterceptor {
+    func performRefresh() async throws {
+        /// If some thread is already refreshing:
+        guard await !refreshingState.isRefreshing else {
+            /// Wait for signal to continue.
+            await refreshingState.wait()
+            await refreshingState.signal()
+            return
+        }
+        
+        defer {
+            Task {
+                /// Unlock refreshing state and signal other threads that refreshing is done.
+                await refreshingState.setIsRefreshing(false)
+                await refreshingState.signal()
+            }
+        }
+        
+        /// Lock refreshing state to prevent other threads from trying to refresh as well.
+        await refreshingState.setIsRefreshing(true)
+        
+        /// Try refreshing authorization data.
+        try await authorizationManager.refreshAuthorizationData()
+    }
+}
+
+// MARK: Private actor
+private extension AuthorizationTokenInterceptor {
+    actor RefreshingState {
+        private let semaphore = AsyncSemaphore(value: 0)
```

Comment:
I try to do the semaphore math and it seems broken to me. Example:
1. Semaphore is instantiated with value 0
2. `performRefresh` is called
3. `isRefreshing` is set to `true`
4. Token is refreshed
5. `isRefreshing` is set to `false`
6. `signal()` is called -> semaphore inner value is increased by one -> semaphore value is 1
7. `performRefresh` is called
8. `isRefreshing` is set to `true`
9. `performRefresh` is called
10. Refreshing is in the process so `wait()` is called -> semaphore inner value is decreased by 1 -> semaphore value is 0 -> value is greater than or equal to 0 -> the task is not suspended and continues -> the task returns from `performRefresh` before the token is refreshed
I believe the correct setup is:
1. Instantiate the semaphore with `value: 1`
2. Call `wait()` when refreshing starts

Reply:
About this I was pinged by @tonyskansf with this [link ](https://www.donnywals.com/building-a-token-refresh-flow-with-async-await-and-swift-concurrency/)
shouldn't we eventually rather use actor instead of semaphore? I haven't read it yet, just opening discussion

Reply:
I vote for changing this to an actor

Reply:
@cejanen @tonyskansf Damn thats a really cool approach 😍, way more straightforward that the semaphore. I will try to implement it.

Reply:
Done ✅ 015087c

Reply:
🎉